provider/internal/queue/queue.go

```diff
@@ -15,144 +17,141 @@ import (
-	// an infinite loop dequeing cid.Undef if we do that.
-	// The provider has it's own select on top of dequeue and will handle this by itself.
-	return nil
+// Len returns the number of CIDs in the queue.
+func (q *Queue) Len() int {
+	return q.q.Len()
 }
 
-// Enqueue puts a cid in the queue
-func (q *Queue) Enqueue(cid cid.Cid) error {
-	select {
-	case q.enqueue <- cid:
-		return nil
-	case <-q.ctx.Done():
-		return errors.New("failed to enqueue CID: shutting down")
-	}
+// Cap returns the queue's capacity.
+func (q *Queue) Cap() int {
+	return q.q.Cap()
 }
 
-// Dequeue returns a channel that if listened to will remove entries from the queue
-func (q *Queue) Dequeue() <-chan cid.Cid {
-	return q.dequeue
+// Close stops the queue and writes remaining data to the datastore.
+func (q *Queue) Close() error {
+	var err error
+	q.closeOnce.Do(func() {
+		if q.ds != nil {
+			// Close queue and wait for remaining are written to datastore.
+			q.q.Close()
+			ctx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
+			defer cancel()
+			err = q.writeRemaining(ctx)
+		} else {
+			// Close and drain queue.
+			q.q.Shutdown()
+		}
+	})
+	return err
 }
 
-// worker run dequeues and enqueues when available.
-func (q *Queue) worker() {
-	var k datastore.Key = datastore.Key{}
-	var c cid.Cid = cid.Undef
-
-	defer q.closed.Done()
-	defer q.close()
-
-	for {
-		if c == cid.Undef {
-			head, err := q.getQueueHead()
+func (q *Queue) readPrevious(ctx context.Context) {
+	qry := query.Query{
+		Orders: []query.Order{query.OrderByKey{}},
+	}
+	results, err := q.ds.Query(ctx, qry)
+	if err != nil {
+		log.Errorw("error querying queue", "err", err)
+		return
+	}
+	defer results.Close()
 
-			switch {
-			case err != nil:
-				log.Errorf("error querying for head of queue: %s, stopping provider", err)
+	qIn := q.q.In()
+	for r := range results.Next() {
+		k := datastore.NewKey(r.Key)
+		c, err := cid.Parse(r.Value)
+		if err != nil {
+			log.Warnf("Error parsing queue entry cid with key (%s), removing it from queue: %s", r.Key, err)
+			if err = q.ds.Delete(ctx, k); err != nil {
+				log.Errorw("Error deleting queue entry", "key", r.Key, "err", err)
 				return
-			case head != nil:
-				k = datastore.NewKey(head.Key)
-				c, err = cid.Parse(head.Value)
-				if err != nil {
-					log.Warnf("error parsing queue entry cid with key (%s), removing it from queue: %s", head.Key, err)
-					err = q.ds.Delete(q.ctx, k)
-					if err != nil {
-						log.Errorf("error deleting queue entry with key (%s), due to error (%s), stopping provider", head.Key, err)
-						return
-					}
-					continue
-				}
-			default:
-				c = cid.Undef
 			}
+			continue
 		}
+		qIn <- c
 
-		// If c != cid.Undef set dequeue and attempt write, otherwise wait for enqueue
-		var dequeue chan cid.Cid
-		if c != cid.Undef {
-			dequeue = q.dequeue
+		if err = q.ds.Delete(ctx, k); err != nil {
+			log.Errorw("Failed to delete queued cid", "cid", c, "key", k, "err", err)
```

Comment:
I would also do the deletions as a batch..

Reply:
Cannot do that du to the way the next queued item is read from the datastore.